### PR TITLE
Feat: Support retrieving modules in private git repo through SSH

### PR DESCRIPTION
**Pull request by motilayo**
This is a PR to Support SSH for retrieving terraform modules in private git repo for this issue: https://github.com/kubevela/terraform-controller/issues/292
There is a corresponding PR in kubevela repo: https://github.com/kubevela/kubevela/pull/5059

This change require a secret which hold the SSH private key and known hosts for the git repo:
known_hosts can be generated using `ssh-keyscan <git-url>`
``` yaml
apiVersion: v1
kind: Secret
metadata:
  name: git-ssh
  namespace: vela-system
type: kubernetes.io/ssh-auth
stringData:
  ssh-privatekey: |
   <SSH Private Key> # the ssh private key used for authenticating git
  known_hosts: |
   <SSH known_hosts>  # use `ssh-keyscan github.com`  to generate known_hosts
```

### CODE REVIEW

File: api/v1beta1/configuration_types.go [hunk 1]
```diff
@@ -67,6 +67,9 @@ type BaseConfigurationSpec struct {
 
 	// Region is cloud provider's region. It will override the region in the region field of ProviderReference
 	Region string `json:"region,omitempty"`
+
+	// GitCredentialsReference specifies the reference to the secret containing the git credentials
+	GitCredentialsReference *types.SecretReference `json:"gitCredentialsReference,omitempty"`
```
Comment (wonderflow) [inline api/v1beta1/configuration_types.go:72]: can we move this field into the ConfigurationSpec outside from `BaseConfigurationSpec`, then it won't affect the CRD in KubeVela repo.

File: api/v1beta1/configuration_types.go [hunk 2]
```diff
@@ -113,6 +116,7 @@ type Backend struct {
 // +kubebuilder:subresource:status
 // +kubebuilder:printcolumn:name="STATE",type="string",JSONPath=".status.apply.state"
 // +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:shortName={conf,velaconf}
```
Comment (wonderflow) [inline api/v1beta1/configuration_types.go:119]:
```suggestion
// +kubebuilder:resource:shortName={conf,terraform-conf}
```

Comment (motilayo) [reply api/v1beta1/configuration_types.go]: done

File: api/v1beta2/configuration_types.go [hunk 3]
```diff
@@ -22,6 +22,7 @@ import (
 
 	state "github.com/oam-dev/terraform-controller/api/types"
 	types "github.com/oam-dev/terraform-controller/api/types/crossplane-runtime"
+	v1 "k8s.io/api/core/v1"
```
Comment (chivalryq) [inline api/v1beta2/configuration_types.go:25]: wrong imports order here, please run `make reviewable`

File: controllers/configuration_controller.go [hunk 4]
```diff
@@ -545,6 +553,18 @@ func (r *ConfigurationReconciler) preCheck(ctx context.Context, configuration *v
 		}
 	}
 
+	gitCreds, err := provider.GetGitCredentialsFromConfiguration(ctx, k8sClient, meta.GitCredentialsReference.Namespace, meta.GitCredentialsReference.Name)
+	if gitCreds == nil {
+		msg := "Git credentials Secret is not found"
+		if err != nil {
+			msg = err.Error()
```
Comment (chivalryq) [inline controllers/configuration_controller.go:560]:
```go
msg=fmt.Sprintf("git credentials Secret not found: %s",err)
```

File: controllers/configuration_controller.go [hunk 5]
```diff
@@ -710,6 +730,10 @@ func (meta *TFConfigurationMeta) assembleTerraformJob(executionType TerraformExe
 			Name:      BackendVolumeName,
 			MountPath: BackendVolumeMountPath,
 		},
+		{
+			Name:      GitAuthConfigVolumeName,
+			MountPath: GitAuthConfigVolumeMountPath,
+		},
```
Comment (chivalryq) [inline controllers/configuration_controller.go:736]: Not all git repo need authentication, please modify all logic that always do it.

Comment (motilayo) [reply controllers/configuration_controller.go]: Done

**Comment by chivalryq** [comment]
Let's add some e2e test. Maybe this repo will helps: https://github.com/isomorphic-git/git-http-mock-server

File: .golangci.yml [hunk 6]
```diff
@@ -179,6 +179,14 @@ issues:
...
         - revive
 
+    - text: "package-comments:"
+      linters:
+        - revive
+    
+    - text: "exported:"
+      linters:
+        - revive
+
```
Comment (chivalryq) [inline .golangci.yml:189]: Why should we add this?

File: controllers/configuration_controller.go [hunk 7]
```diff
@@ -879,8 +892,12 @@ func (meta *TFConfigurationMeta) assembleExecutorVolumes() []v1.Volume {
 	workingVolume.EmptyDir = &v1.EmptyDirVolumeSource{}
 	inputTFConfigurationVolume := meta.createConfigurationVolume()
 	tfBackendVolume := meta.createTFBackendVolume()
-	gitAuthConfigVolume := meta.createGitAuthConfigVolume()
-	return []v1.Volume{workingVolume, inputTFConfigurationVolume, tfBackendVolume, gitAuthConfigVolume}
+	if meta.GitCredentialsReference != nil {
+		gitAuthConfigVolume := meta.createGitAuthConfigVolume()
+		return []v1.Volume{workingVolume, inputTFConfigurationVolume, tfBackendVolume, gitAuthConfigVolume}
```
Comment (chivalryq) [inline controllers/configuration_controller.go:897]: Please pre-define a slice and keep one `return` in function

File: e2e/normal/configuration_test.go [hunk 8]
```diff
@@ -284,6 +287,80 @@ func TestForceDeleteConfiguration(t *testing.T) {
...
+		OutputsSecretName:      "random-e2e-git-creds-secret-ref-conn",
+		VariableSecretName:     "variable-random-e2e-git-creds-secret-ref",
+	}
+
+	clientSet, err := client.Init()
+	assert.NilError(t, err)
+	pwd, _ := os.Getwd()
+	gitServer := filepath.Join(pwd, "..", "../examples/git-credentials")
+	gitServerApplyCmd := fmt.Sprintf("kubectl apply -f %s", gitServer)
+	gitServerDeleteCmd := fmt.Sprintf("kubectl apply -f %s", gitServer)
```
Comment (chivalryq) [inline e2e/normal/configuration_test.go:305]: `kubectl delete -f` ?

File: examples/git-credentials/git-ssh-auth-secret.yaml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
```
Comment (chivalryq) [inline examples/git-credentials/git-ssh-auth-secret.yaml:1]: Seems like this is a intermedium generated file? If so, it shouldn't be commited.

File: examples/git-credentials/git-push-job.yaml [hunk 10]
```diff
@@ -0,0 +1,86 @@
+apiVersion: v1
```
Comment (chivalryq) [inline examples/git-credentials/git-push-job.yaml:1]: What's the purpose of this file?

File: examples/git-credentials/ssh-keys-secret.yaml [hunk 11]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
```
Comment (chivalryq) [inline examples/git-credentials/ssh-keys-secret.yaml:1]: And this file?

**Comment by chivalryq** [comment]
@motilayo You did a very great job! The test methods is impressive. And can you explain some of the files which is not used in test?
Don't worry about the code coverage CI item. It seems not to calculate the e2e test correctly.

File: .golangci.yml (hunk 6 above)
Comment (motilayo) [reply .golangci.yml:189]: The address linting errors, in [ci lint](https://github.com/kubevela/terraform-controller/actions/runs/3560027171/jobs/5979782813#step:4:86)
```
Error: controllers/configuration/configuration.go:1:1: package-comments: should have a package comment (revive)
```

```
Error: api/types/state.go:43:2: exported: exported const TerraformInit should have comment (or a comment on this block) or be unexported (revive)
```

File: examples/git-credentials/git-push-job.yaml (hunk 10 above)
Comment (motilayo) [reply examples/git-credentials/git-push-job.yaml:1]: This job will push to the git-server which is running in a different pod.

File: examples/git-credentials/ssh-keys-secret.yaml (hunk 11 above)
Comment (motilayo) [reply examples/git-credentials/ssh-keys-secret.yaml:1]: That's the secret file for the ssh public key & private key. The public key is added to the git-server pod, the private key is used to push to the git server by the git-push job.

**Comment by motilayo** [comment]
Those files in `examples/git-credentials` are prereqs for support git ssh credentials test. The git-server which requires ssh auth for git push & pull. The git-push job which will push the terraform module to a repo on the git server pod, the secret with the ssh public key & private key. The public key is added to authorized_keys file on the git server, the private key is added to the git-push for the initial repo push. Then a generated secret which holds the private key & known_hosts - this secret is mounted to the terraform apply/destroy job so that it can pull the terraform module from a private repo.
Best effort to keep everything self contained for e2e test.

File: .golangci.yml (hunk 6 above)
Comment (chivalryq) [reply .golangci.yml:189]: OK, got it.

**Comment by motilayo** [comment]
Hi @chivalryq & @wonderflow , hoping for a review. I've made the suggested changes and added some tests. Let me know if there's any pending tasks for me to do.
Thanks!

File: .github/workflows/e2e-test.yml [hunk 12]
```diff
@@ -44,6 +44,7 @@ jobs:
           kind create cluster --image kindest/node:v1.20.7@sha256:688fba5ce6b825be62a7c7fe1415b35da2bdfbb5a69227c499ea4cc0008661ca
           kubectl version
           kubectl cluster-info
+          kubectl describe nodes
```
Comment (chivalryq) [inline .github/workflows/e2e-test.yml:47]: Why should we change this?

File: controllers/configuration_controller.go [hunk 13]
```diff
@@ -1279,3 +1331,32 @@ func (meta *TFConfigurationMeta) RenderConfiguration(configuration *v1beta2.Conf
...
+	if _, ok := secret.Data[v1.SSHAuthPrivateKey]; !ok {
+		errMsg := fmt.Sprintf("'%s' not in git credentials secret", v1.SSHAuthPrivateKey)
+		klog.ErrorS(keyNotFoundErr, errMsg, "Name", secretRef.Name, "Namespace", secretRef.Namespace)
+		return nil, errors.Wrap(keyNotFoundErr, errMsg)
+	}
+	if _, ok := secret.Data["known_hosts"]; !ok {
+		errMsg := "'known_hosts' not in git credentials secret"
+		klog.ErrorS(keyNotFoundErr, errMsg, "Name", secretRef.Name, "Namespace", secretRef.Namespace)
+		return nil, errors.Wrap(keyNotFoundErr, errMsg)
+	}
```
Comment (chivalryq) [inline controllers/configuration_controller.go:1359]: 1. Please make "known_hosts" a const string, you can both use it in controller and test.
2. Make the detect a loop like
```go
needSecretKeys:=[]string{v1.SSHAuthPrivateKey, GitCredsKnownHosts}
for _,key:=range needSecretKeys{
}
```
3. reduce the message wrap, just return `errors.Errorf("%s key not found in git credentials secret", key) `is ok

File: controllers/configuration_controller_test.go [hunk 14]
```diff
@@ -2497,3 +2531,134 @@ terraform {
...
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			got, err := GetGitCredentialsSecret(ctx, tc.args.k8sClient, tc.args.GitCredentialsSecretReference)
+
+			if err != nil {
+				if !strings.Contains(err.Error(), tc.want.errMsg) {
+					t.Errorf("GetGitCredentialsSecret error = %v, wantErr = %v", err, tc.want.errMsg)
+					return
+				}
```
Comment (chivalryq) [inline controllers/configuration_controller_test.go:2657]: use `assert.Contains(t, ...)` is simpler

**Review by chivalryq — APPROVED** [review]
LGTM after some code cleanup

File: .github/workflows/e2e-test.yml (hunk 12 above)
Comment (motilayo) [reply .github/workflows/e2e-test.yml:47]: not needed, was using it to debug earlier e2e issue. Will revert it

File: controllers/configuration_controller.go (hunk 13 above)
Comment (motilayo) [reply controllers/configuration_controller.go]: done

File: controllers/configuration_controller_test.go (hunk 14 above)
Comment (motilayo) [reply controllers/configuration_controller_test.go]: done

**Comment by chivalryq** [comment]
@motilayo You can fix the conflicts within PR in vela's repo and we try to merge it into master.

**Pull request merged**